### PR TITLE
coredump: add bits to enable dumping privilege stack during coredump

### LINKED ISSUE #78421: Capture privileged stack in coredump when userspace is enabled, so that coredumps occuring in kernel mode contain relevant stack info
**Is your enhancement proposal related to a problem? Please describe.**

> Note that the following applies to Xtensa. I have not verified if the same problem exists in, or is applicable to, other architectures. 

When collecting a coredump, the stack is captured and used to produce a backtrace. Currently only the privileged region of the stack is captured. Thus, any portion of the backtrace that occurs in kernel mode is not available. 

image

**Describe the solution you'd like**
I think that we should add some fields to the `stack_info` structure, to specify the start and size of the privileged portion of the stack. Alternately, we could make this an architecture-specific callback in the coredump collection process, but that might be a more complicated implementation.

### PR DESCRIPTION
**Pull request by dcpleung**
See individual commits for details.

Only RISC-V, x86 and Xtensa are supported right now as they are simple enough to figure out where the privilege stacks are.

### CODE REVIEW

File: arch/Kconfig [hunk 1]
```diff
@@ -114,6 +115,7 @@ config RISCV
 	bool
 	select ARCH_IS_SET
 	select ARCH_SUPPORTS_COREDUMP
+	select ARCH_SUPPORTS_COREDUMP_PRIV_STACKS
```
Comment (agiardin) [inline arch/Kconfig:118]: `select ARCH_SUPPORTS_COREDUMP_PRIV_STACKS` needs to also be added to `config XTENSA` right?

File: subsys/debug/coredump/Kconfig [hunk 2]
```diff
@@ -107,4 +107,15 @@ config DEBUG_COREDUMP_THREADS_METADATA
...
 
+config DEBUG_COREDUMP_DUMP_THREAD_PRIV_STACK
+	bool "Dump privilege stack of user threads"
+	default y
+	depends on ARCH_SUPPORTS_COREDUMP_PRIV_STACKS
+	depends on USERSPACE
+	help
+	  Dump the privilege stack of user threads.
+
+	  Say n to conserve space on coredump backend or that you will never
```
Comment (agiardin) [inline subsys/debug/coredump/Kconfig:118]:
```suggestion
	  Say n to conserve space on coredump backend or if you will never
```

Comment (dcpleung) [reply subsys/debug/coredump/Kconfig]: Done.

File: arch/Kconfig (hunk 1 above)
Comment (dcpleung) [reply arch/Kconfig:118]: On Xtensa, coredump is only enabled by the SoC or board. So this is not done via arch kconfig.

Comment (dcpleung) [reply arch/Kconfig:118]: I have moved the kconfig to the SoC instead of board, which is more appropriate.

File: arch/riscv/core/coredump.c [hunk 3]
```diff
@@ -116,3 +117,21 @@ uint16_t arch_coredump_tgt_code_get(void)
 {
 	return COREDUMP_TGT_RISC_V;
 }
+
+#if defined(CONFIG_DEBUG_COREDUMP_DUMP_THREAD_PRIV_STACK)
+void arch_coredump_priv_stack_dump(struct k_thread *thread)
```
Comment (ycsin) [inline arch/riscv/core/coredump.c:122]: There's a similar implementation in `stacktrace.c`

https://github.com/zephyrproject-rtos/zephyr/blob/74c5c84c66d59320ab39b58662d60e6cb6f188c2/arch/riscv/core/stacktrace.c#L55-L71

Comment (dcpleung) [reply arch/riscv/core/coredump.c:122]: Thanks. Updated.

**Comment by dcpleung** [comment]
Twister failure is due to: https://github.com/zephyrproject-rtos/zephyr/issues/78677

File: tests/subsys/debug/coredump/testcase.yaml [hunk 4]
```diff
@@ -22,3 +22,31 @@ tests:
         - "E: #CD:4([dD])([0-9a-fA-F]+)"
         - "E: #CD:END#"
         - "k_sys_fatal_error_handler"
+  debug.coredump.logging_backend.usespace:
```
Comment (mrkhldn) [inline tests/subsys/debug/coredump/testcase.yaml:25]: debug.coredump.logging_backend.userspace

File: tests/subsys/debug/coredump/testcase.yaml [hunk 5]
```diff
@@ -22,3 +22,31 @@ tests:
...
+    harness_config:
+      type: multi_line
+      regex:
+        - "Coredump: (.*)"
+        - ">>> ZEPHYR FATAL ERROR "
+        - "E: #CD:BEGIN#"
+        - "E: #CD:5([aA])45([0-9a-fA-F]+)"
+        - "E: #CD:41([0-9a-fA-F]+)"
+        - "E: #CD:4([dD])([0-9a-fA-F]+)"
+        - "E: #CD:4([dD])([0-9a-fA-F]+)"
```
Comment (mrkhldn) [inline tests/subsys/debug/coredump/testcase.yaml:50]: Should there be at least one more line of the following from the call to arch_coredump_priv_stack_dump

"E: #CD:4([dD])([0-9a-fA-F]+)"

**Review by mrkhldn — DISMISSED** [review]

File: tests/subsys/debug/coredump/testcase.yaml (hunk 4 above)
Comment (dcpleung) [reply tests/subsys/debug/coredump/testcase.yaml]: Oops... fixed.

File: tests/subsys/debug/coredump/testcase.yaml (hunk 5 above)
Comment (dcpleung) [reply tests/subsys/debug/coredump/testcase.yaml:50]: Yes... added.

**Review by mrkhldn — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by ycsin — APPROVED** [review]

**Pull request merged**